substrate/client/consensus/beefy/src/communication/fisherman.rs

```diff
@@ -0,0 +1,396 @@
+};
+use sp_mmr_primitives::{AncestryProof, MmrApi};
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block, Header, NumberFor},
+};
+use std::{marker::PhantomData, sync::Arc};
+
+/// Helper wrapper used to check gossiped votes for (historical) equivocations,
+/// and report any such protocol infringements.
+pub(crate) struct Fisherman<B: Block, BE, P, R> {
+	pub backend: Arc<BE>,
+	pub runtime: Arc<R>,
+	pub key_store: Arc<BeefyKeystore<AuthorityId>>,
+	pub payload_provider: P,
+	pub _phantom: PhantomData<B>,
+}
+
+struct CanonicalHashHeaderPayload<B: Block> {
+	hash: B::Hash,
+	header: B::Header,
+	payload: Payload,
+}
+
+impl<B, BE, R, P> Fisherman<B, BE, P, R>
+where
+	B: Block,
+	BE: Backend<B> + Send + Sync,
+	P: PayloadProvider<B> + Send + Sync,
+	R: ProvideRuntimeApi<B> + Send + Sync,
+	R::Api: BeefyApi<B, AuthorityId, MmrRootHash> + MmrApi<B, MmrRootHash, NumberFor<B>>,
+{
+	fn canonical_hash_header_payload(
+		&self,
+		number: NumberFor<B>,
+	) -> Result<CanonicalHashHeaderPayload<B>, Error> {
+		// This should be un-ambiguous since `number` is finalized.
+		let hash = self
+			.backend
+			.blockchain()
+			.expect_block_hash_from_id(&BlockId::Number(number))
+			.map_err(|e| Error::Backend(e.to_string()))?;
+		let header = self
+			.backend
+			.blockchain()
+			.expect_header(hash)
+			.map_err(|e| Error::Backend(e.to_string()))?;
+		self.payload_provider
+			.payload(&header)
+			.map(|payload| CanonicalHashHeaderPayload { hash, header, payload })
+			.ok_or_else(|| Error::Backend("BEEFY Payload not found".into()))
+	}
+
+	fn active_validator_set_at(
+		&self,
+		block_hash: <<B as Block>::Header as Header>::Hash,
+	) -> Result<ValidatorSet<AuthorityId>, Error> {
+		let header = self
+			.backend
+			.blockchain()
+			.expect_header(block_hash)
+			.map_err(|e| Error::Backend(e.to_string()))?;
+		expect_validator_set_nonblocking(&*self.runtime, &*self.backend, &header)
+			.map_err(|e| Error::Backend(e.to_string()))
+	}
+
+	pub(crate) fn report_fork_equivocation(
+		&self,
+		proof: ForkEquivocationProof<AuthorityId, B::Header, MmrRootHash>,
+	) -> Result<bool, Error> {
+		let best_block_number = self.backend.blockchain().info().best_number;
+		let best_block_hash = self.backend.blockchain().info().best_hash;
+
+		// if the commitment is for a block number exceeding our best block number, we assume the
+		// equivocators are part of the current validator set, hence we use the validator set at the
+		// best block
+		let canonical_commitment_block_hash = if best_block_number < proof.commitment.block_number {
+			best_block_hash
+		} else {
+			self.backend
+				.blockchain()
+				.expect_block_hash_from_id(&BlockId::Number(proof.commitment.block_number))
+				.map_err(|e| Error::Backend(e.to_string()))?
+		};
+
+		let validator_set = self.active_validator_set_at(canonical_commitment_block_hash)?;
+		let set_id = validator_set.id();
+
+		let best_mmr_root = self
+			.runtime
+			.runtime_api()
+			.mmr_root(best_block_hash)
```

Comment:
The `best_mmr_root` here is what's ultimately passed down as the `root` arg of [`sp_merkle_mountain_range::utils::verify_ancestry_proof(root, mmr_size, ancestry_proof)`](https://github.com/lederstrumpf/polkadot-sdk/blob/d8625d8bdae36b71e78394613ec00264b1159f3f/substrate/primitives/merkle-mountain-range/src/utils.rs#L46-L50), i.e. it's the current mmr root, which the ancestry proof proves descends from some `prev_root` @ `commitment.block_number` where `prev_root != commitment.payload`.

The `canonical_mmr_root` you're suggesting is in fact implicit in the ancestry proof: it's calculated also calculated in `sp_merkle_mountain_range::utils::verify_ancestry_proof(root, mmr_size, ancestry_proof)`: https://github.com/lederstrumpf/polkadot-sdk/blob/d8625d8bdae36b71e78394613ec00264b1159f3f/substrate/primitives/consensus/beefy/src/lib.rs#L494-L499.
This in fact has to be calculated inside the proof rather than fed in externally since the runtime can't look up past mmr roots (else, our lives would be *much* simpler 😄).

So the original code over here is correct, but I'm happy you pointed me towards this because I used conflated senses of "canonical" downstream from this - see d8625d8bdae36b71e78394613ec00264b1159f3f for the renaming changes: should be clearer to follow now.

Reply:
On a related note: gentle prod for feedback on [suggestion](https://github.com/paritytech/polkadot-sdk/pull/1903#issue-1946874074:~:text=B.%20alternative%20implementation) to change the handling to not use the best mmr root, but instead requiring the prover to always provide a header in the proof, and then proving ancestry of the correct `prev_root` against the root of said header.
The implementation of this is more complicated than the current one:
1. if no ancestry proof is provided, the provided header is the canonical header at height `commitment.block_number`, and
2. if an ancestry proof is provided, the provided header is a recent header whose hash can still be verified onchain and whose mmr root descends from the correct `prev_root`)

But this change would avoid the current requirement that the fork equivocation proof is processed in the same block it was generated for, so allowing it to idle -- say, on the mempool -- before being processed (in current impl, it can only be successfully submitted by the block producer).

Reply:
I think it is good enough to be submitted by the block producer - especially since block producers are implicitly running the BEEFY fisherman - so as long as there's at least one honest validator/block-producer it should be included.

Reply:
I think it would be good to accommodate the case where the equivocation is not submitted by the block producer. It could be summited for example by a bridge

Reply:
It would be good yes, better even. BUT is it critical? I am partial to a simple but working solution vs a comprehensive but complex one.

What are the costs vs benefits?

Reply:
In practice, validators looking to maximize economic gains will always "steal" any equivocation reporting (which is a profitable action) and include it themselves as the block producer.
In practice, external (non-validator) fishermen operate at a loss and there is low incentive for their existence (when validators also act as fishermen).

Reply:
Ok, I see. Anyway high level the solution seems really simple. We just need to add one header. I won't make it a priority, but I will give it a try if I have time.